### ISSUE TITLE
No way to verify that a list argument's entries are valid in a const assert

### ISSUE THREAD

**Issue by Hixie** [dart-lang/sdk#29276]
```dart
abstract class MultiChildRenderObjectWidget extends RenderObjectWidget {
  const MultiChildRenderObjectWidget({ Key key, this.children: const <Widget>[] })
    : assert(children != null),
      assert(!children.any((Widget child) => child == null)), // <-- this line won't compile
      super(key: key);
  // ...
}
```

**Comment by lrhn** [comment]
Correct. Const expressions cannot iterate over lists or call methods. That is not new, it is a know restriction on compile-time constant expressions, and not something we expect to change.

That does restrict what you can do with compile-time constant asserts, but no more than it restricts what you can do with const expressions in general.

As a rule, you cannot deconstruct composite values (whether lists or other objects) at compile time, you can only build them.

**Comment by mathieujobin** [comment]
`assert(children.all((Widget child) => child != null))` would be more readable IMO

**Comment by tvolkert** [comment]
Revisiting this.  This one assert is too useful to remove, and the fact that it's keeping `MultiChildRenderObjectWidget` from being const constructible is keeping large swaths of app code from being const constructible (because `Row` and `Column` are both descendants of `MultiChildRenderObjectWidget`).

**Comment by dnfield** [comment]
When non-nullability comes in, will we have `List<T>` where `T` is non-nullable?

**Comment by eernstg** [comment]
Definitely yes.

**Comment by a14n** [comment]
Plugged question regarding assert in initializer list: I often wonder why they are part of the const evaluation? If they were executed only at the first initialization we could make almost every widget classes const (especially Row,  Column, and Stack) without changes.

**Comment by cedvdb** [comment]
How is that assert valuable ? If the children is a List<Widget> the elements cannot be null, can they ?

**Comment by dnfield** [comment]
It's still valuable as long as we have customers running in unsound null-safety mode, and we still have a substantial number of those.

**Comment by gustav3d** [comment]
There are other breaking changes happening all the time.
So no,  it is not,  it's their problem that they refuse to fix their code
but still expect things to function in any new version.
So let them be stuck on version X like in all other cases of breaking
changes.

**Comment by cedvdb** [comment]
I did not know they could run on unsound null safety while at the same time receive updates from flutter.. That sounds like a significant maintenance annoyance

**Comment by mateusfccp** [comment]
When will unsound null-safe mode be deprecated?

We shouldn't restrain ourselves for so much time because some people refuses to progress...

**Comment by lrhn** [comment]
Don't restrain yourself. Write perfectly good null safe code, then allow that code to crash and burn if someone passes in a `null` argument to a non-nullable parameter. Don't bother writing `if (arg as dynamic == null) throw NiceError("this thing shouldn't be null")`. Just let the dice roll and see where the `null` crashes things.

They had it coming. They only had themselves to blame.

**Comment by dnfield** [comment]
The last I checked on what data we have for this, >20% of Flutter users are using unsound null-safety. As nice as it would be to just migrate, the user experience of getting nonsensical errors at some unexpected place in your program is pretty bad.

**Comment by larssn** [comment]
Isn't this one of the things that's purely theoretical? Do you have any data on how often people run into this assertion?

For us, still running unsound, we've never run into it.

It seems like a pretty niche thing, and for it to hold the codebase back for the majority of users is not something I agree with, even though we're not going to benefit from it atm.

So pull the trigger. 🙂

**Comment by dnfield** [comment]
> Isn't this one of the things that's purely theoretical? Do you have any data on how often people run into this assertion?
> 
> For us, still running unsound, we've never run into it.
> 
> It seems like a pretty niche thing, and for it to hold the codebase back for the majority of users is not something I agree with, even though we're not going to benefit from it atm.
> 
> So pull the trigger. 🙂

We have data that a large number of users are still running code in unsound null-safe mode, where this assertion is helpful. 

Dart 3 will be removing this problem before very long. We can just wait for that.
